include/zephyr/drivers/adc.h

```diff
@@ -704,9 +712,27 @@ __subsystem struct adc_driver_api {
 #ifdef CONFIG_ADC_ASYNC
 	adc_api_read_async    read_async;
 #endif
+	adc_api_get_ref_internal get_ref_internal;
 	uint16_t ref_internal;	/* mV */
 };
 
+/**
+ * @brief Verify that a device has an ADC driver API.
+ *
+ * @param dev a device pointer
+ *
+ * @return true if the device's API is located in the ADC API section.
+ */
+static inline bool device_has_adc_api(const struct device *dev)
+{
+	STRUCT_SECTION_START_EXTERN(adc_driver_api);
+	STRUCT_SECTION_END_EXTERN(adc_driver_api);
+
+	return IN_RANGE((const struct adc_driver_api *)dev->api,
+			STRUCT_SECTION_START(adc_driver_api),
+			STRUCT_SECTION_END(adc_driver_api) - 1);
+}
```

Comment:
This allows us to do:

```c
	STRUCT_SECTION_FOREACH(device, dev) {
		if (device_has_adc_api(dev)) {
			/* do somethin ADC related */
		}
	}
```

Reply:
Maybe `device_has_adc_driver_api` in full is better?

Reply:
The issue is that the `extern` symbols need to be declared first.

Reply:
Yeah I think the foreach could also be done, something similar to how `STRUCT_SECTION_FOREACH` itself is implemented but instead of just incrementing the iterator this could use a helper function that takes a pointer to the iterator, the boundaries of the API and the end of the struct device as an argument, and then it increments the iterator while performing the check to skip the devices we don't care about (and not go beyond the device section). Yikes. :-)

Reply:
I've added your `DRIVER_API_IS` and `DRIVER_API_EXTERN` to keep everything even shorter :)

Reply:
> The issue is that the `extern` symbols need to be declared first.

you can wrap local variables in a pair of `{}` to keep them in the context of the macro expansion and still return a value, check out https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/util.h#L827C9-L827C17

Reply:
I guess 😅 

But I think this is easily covered with the `FOREACH` device..?

Reply:
Also keep in mind we can get the API from a device, but not the other way around.

Reply:
@fabiobaltieri I think I figured out what you meant, please take a look 😄

Reply:
Yes that's exactly what I meant, looks quite neat now :-) I'll have a go at implementing a generic foreach, this is getting very zephyresque at this point